src/protocol/sort/bit_permutations.rs

```diff
@@ -0,0 +1,154 @@
+use std::marker::PhantomData;
+
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{
+        mesh::{Gateway, Mesh},
+        prss::SpaceIndex,
+    },
+    protocol::{context::ProtocolContext, IPAProtocolStep, RecordId, SortStep, Step},
+    secret_sharing::Replicated,
+};
+
+use futures::future::join_all;
+
+#[derive(Debug)]
+pub struct BitPermutations<'a, G, M, S: Step + SpaceIndex> {
+    pub(crate) context: &'a ProtocolContext<'a, G, S>,
+    pub(crate) phantom: PhantomData<M>,
+}
+
+impl<'a, G: Gateway<M, IPAProtocolStep>, M> BitPermutations<'a, G, M, IPAProtocolStep>
+where
+    M: Mesh,
+{
+    /// In this step, multiplication inputs are generated locally at each helper from input share, x in following steps
+    /// 1. calculate 1 - x, x and concatenate them
+    /// 2. calculate cumulative sum at each vector row
+    /// 3. return back tuple of step 1 and step 2 output
+    fn prepare_mult_inputs<F>(
+        &self,
+        shares: &[Replicated<F>],
+    ) -> Vec<(Replicated<F>, Replicated<F>)>
+    where
+        F: Field,
+    {
+        let channel = self
+            .context
+            .gateway
+            .get_channel(IPAProtocolStep::Sort(SortStep::BitPermutations));
+
+        shares
+            .iter()
+            .map(|x| channel.share_of_one() - *x)
+            .chain(shares.iter().copied())
+            .scan(Replicated::<F>::new(F::ZERO, F::ZERO), |sum, n| {
+                *sum = *sum + n;
+                Some((n, *sum))
+            })
+            .collect()
+    }
+
+    /// multiplies the input vector pairs across helpers and returns result
+    /// For this, it spawns all multiplication, wait for them to finish in parallel and then collect the results
+
+    #[allow(clippy::cast_possible_truncation)]
+    async fn secure_multiply<F>(
+        &self,
+        mult_inputs: Vec<(Replicated<F>, Replicated<F>)>,
+    ) -> Vec<Replicated<F>>
```

Comment:
I think as it involves communication, it is fallible, therefore `Result<Vec<_>, E>` is more appropriate